public/app/plugins/datasource/stackdriver/components/Metrics.tsx

```diff
@@ -0,0 +1,192 @@
+import React from 'react';
+import _ from 'lodash';
+
+import appEvents from 'app/core/app_events';
+import { MetricSelect } from 'app/core/components/Select/MetricSelect';
+
+export interface Props {
+  onChange: (metricDescriptor) => void;
+  templateSrv: any;
+  datasource: any;
+  defaultProject: string;
+  metricType: string;
+  children?: (renderProps: any) => JSX.Element;
+}
+
+interface State {
+  metricDescriptors: any[];
+  metrics: any[];
+  services: any[];
+  service: string;
+  metric: string;
+  metricDescriptor: any;
+  defaultProject: string;
+}
+
+export class Metrics extends React.Component<Props, State> {
+  state: State = {
+    metricDescriptors: [],
+    metrics: [],
+    services: [],
+    service: '',
+    metric: '',
+    metricDescriptor: null,
+    defaultProject: '',
+  };
+
+  constructor(props) {
+    super(props);
+  }
+
+  componentDidMount() {
+    this.setState({ defaultProject: this.props.defaultProject }, () => {
+      this.getCurrentProject()
+        .then(this.loadMetricDescriptors.bind(this))
+        .then(this.initializeServiceAndMetrics.bind(this));
+    });
+  }
+
+  async getCurrentProject() {
+    return new Promise(async (resolve, reject) => {
```

Comment:
this looks strange, in every function (and creation of new promise in each), looks like the error handling could be done in componentDidMount and getCurrentProject & loadMetricDescriptors & initializeServiceAndMetrics can be simplified a bit (no need to do that promise creation and error handling)